Caltech-IPAC/firefly
# DM-6516-footprint footprint tool implementation. Integrating with the  implementation of marker tool. Both are made in consistent style.

Implementation: (both footprint tool and marker tool are made in consistent style)   
- footprint description file - in DS9 region format, footprint description is located at the server side.
-  update the service to load the footprint description from the server. 
-  create footprint factory to produce regions and drawing objects based on footprint description. 
-  create  drawing layer for footprint.
-  create functions to handle the operation on footprint (or marker) on the drawing layer - rotate, resize, or relocate or translation.
-  handle the cases that the footprint coverage is bigger than the image size or the footprint's position is out of image covered area.          
- UI components for drawing layer panel

Notes for operation: 
Rotate handle, resize handle and outline box are provided for the user to rotate, resize and relocate the footprint or marker on the U if any of the operations is applied.
- Outline box is initially an upright rectangle which
    a) encloses all footprint regions or
    b) is around the center of the footprint (marked by a cross symbol on the UI) or the marker if a) is not visible or 
    c) is around the plot center if none of a) and  b) is visible.  
    The outline box becomes slanted as the footprint is rotated. 
- Resize handles appear at the corners of the outline box. (marker is resizable and footprint is not resizable).
- Rotate handle appears only at one side of the outline box, starting from the right, the top, the left, or the bottom side  according to its visibility on the plot.
  
Testing example,
   1. get image from image search panel (Images), target: m51, select WISE and click search button
   2. click icon for "overlay marker and instrument footprints"
      and select "add marker" or any item for adding footprint from the dropdown list
      => a marker or the footprint will appear in the middle of the plot
   3. click the mouse at the place where you want to locate the footprint           

   When the mouse hovers over the footprint, the cursor will change to be 'pointer'.
   Once 'pointer' is clicked down, an outline box with either resize or rotate handle will appear for 3 seconds after the mouse is up.
  
   4. Click the resize handle and drag to resize the marker,  or 
       click the rotate handle and drag to rotate the footprint or
       click on the pointer and drag to move the marker or footprint. 

   If the outline box is moved out of the plot area  while operating the footprint or marker, the outline box of other kind will continue to appear.
